src/storage/AddEdgesProcessor.cpp

```diff
@@ -13,27 +13,133 @@ namespace nebula {
+            auto partId = partEdges.first;
+            std::vector<kvstore::KV> data;
+            std::for_each(partEdges.second.begin(), partEdges.second.end(), [&](auto& edge){
+                VLOG(4) << "PartitionID: " << partId << ", VertexID: " << edge.key.src
+                        << ", EdgeType: " << edge.key.edge_type << ", EdgeRanking: "
+                        << edge.key.ranking << ", VertexID: "
+                        << edge.key.dst << ", EdgeVersion: " << version;
+                auto key = NebulaKeyUtils::edgeKey(partId, edge.key.src, edge.key.edge_type,
+                                                   edge.key.ranking, edge.key.dst, version);
+                data.emplace_back(std::move(key), std::move(edge.get_props()));
+            });
+            doPut(spaceId_, partId, std::move(data));
         });
-        doPut(spaceId, partId, std::move(data));
+    } else {
+        std::for_each(req.parts.begin(), req.parts.end(), [&](auto& partEdges){
+            auto partId = partEdges.first;
+            const auto &edges = partEdges.second;
+            auto atomic = [&]() -> std::string {
+                return addEdges(version, partId, edges);
+            };
+            auto callback = [partId, this](kvstore::ResultCode code) {
+                handleAsync(spaceId_, partId, code);
+            };
+            this->kvstore_->asyncAtomicOp(spaceId_, partId, atomic, callback);
+        });
+    }
+}
+
+std::string AddEdgesProcessor::addEdges(int64_t version, PartitionID partId,
+                                        const std::vector<cpp2::Edge>& edges) {
+    std::unique_ptr<kvstore::BatchHolder> batchHolder = std::make_unique<kvstore::BatchHolder>();
+    std::map<std::string, std::pair<std::string,
+             std::pair<cpp2::EdgeKey, cpp2::IndexItem>>> newIndexes;
+    std::for_each(edges.begin(), edges.end(), [&](auto& edge){
+        auto prop = edge.get_props();
+        auto type = edge.key.edge_type;
+        auto srcId = edge.key.src;
+        auto rank = edge.key.ranking;
+        auto dstId = edge.key.dst;
+        VLOG(4) << "PartitionID: " << partId << ", VertexID: " << srcId
+                << ", EdgeType: " << type << ", EdgeRanking: " << rank
+                << ", VertexID: " << dstId << ", EdgeVersion: " << version;
+        auto key = NebulaKeyUtils::edgeKey(partId, srcId, type, rank, dstId, version);
+        batchHolder->put(std::move(key), std::move(prop));
+        for (auto& index : indexes_) {
+            /**
+             * skip the in-edge.
+             **/
+            if (index.get_schema() == type) {
+                cpp2::EdgeKey ek(apache::thrift::FragileConstructor::FRAGILE,
+                        srcId, type, rank, dstId);
+                auto ekVal = NebulaKeyUtils::edgePrefix(partId, srcId, type, rank, dstId);
+                newIndexes[ekVal] = std::make_pair(edge.get_props(), std::make_pair(ek, index));
```

Comment:
> Why we encode the edgeKey both in key and value of `newIndexes`? I guess you mean to prevent from different edge of same property value? Maybe `map<propertry, vector<edgekey>>` is enough?

good point ！There are two points here：
1: As your said,  different edge of same property value.
2: Same edge but different version number in a request.

Reply:
Still a little confused, I have not see the place we use the key of `newIndexes` (aka `ekVal`). Besides, `ek` and `ekVal` is almost the same.

As the second case `Same edge but different version number in a request`, we didn't encode any version here. I see you insert a new index in `newIndex` and discards all index of the edges we insert in `obsoleteIndex`.

Reply:
> Still a little confused, I have not see the place we use the key of `newIndexes` (aka `ekVal`). Besides, `ek` and `ekVal` is almost the same.
> 
> As the second case `Same edge but different version number in a request`, we didn't encode any version here. I see you insert a new index in `newIndex` and discards all index of the edges we insert in `obsoleteIndex`.

For example : 
insert below edges in one request:
kv (part1_src1_edgeType1_rank1_dst1 , v1)
kv (part1_src1_edgeType1_rank1_dst1 , v2)
kv (part1_src1_edgeType1_rank1_dst1 , v3)
kv (part1_src1_edgeType1_rank1_dst1 , v4)

Ultimately, only one record is stored in newIndexes ：kv (part1_src1_edgeType1_rank1_dst1 , v4) . It's just what I need.Answer your question，the `ekVal` only is a unique key.

Reply:
👍